pkg/receive/writer.go

```diff
@@ -41,17 +45,26 @@ func (r *Writer) Receive(wreq *prompb.WriteRequest) error {
 			}
 		}
 
+		// Append as many valid samples as possible, but keep track of the errors
 		for _, s := range t.Samples {
 			_, err = app.Add(lset, s.Timestamp, s.Value)
-			if err != nil {
-				return errors.Wrap(err, "failed to non-fast add")
+			switch err {
+			case nil:
+				continue
+			case storage.ErrOutOfOrderSample:
+				level.Debug(r.logger).Log("msg", "Out of order sample", "lset", lset.String(), "sample", s.String())
+			case storage.ErrDuplicateSampleForTimestamp:
+				level.Debug(r.logger).Log("msg", "Duplicate sample for timestamp", "lset", lset.String(), "sample", s.String())
+			case storage.ErrOutOfBounds:
+				level.Debug(r.logger).Log("msg", "Out of bounds metric", "lset", lset.String(), "sample", s.String())
 			}
+			errs.Add(errors.Wrap(err, "failed to non-fast add"))
 		}
 	}
 
 	if err := app.Commit(); err != nil {
```

Comment:
if the errs slice is length > 0 then we should never even try to commit, no?

Reply:
ping @dtrejod

Reply:
We don't want to rollback. Per discussion before -- https://github.com/thanos-io/thanos/pull/1563#discussion_r329066801 -- we should always commit all valid samples. We simply should ignore the samples in conflict and continue on processing.